index.ts

```diff
@@ -269,85 +271,94 @@ export default class Analytics {
   /**
    * Flushes the message queue to the server immediately if a flush is not already in progress.
    */
-  flush(callback: AnalyticsFlushCallback = () => {}): void {
-    // check if earlier flush was pushed to queue
-    this.logger.debug('in flush');
-
-    if (!this.enable) {
-      setImmediate(callback);
-      return;
-    }
-
-    if (this.timer) {
-      this.logger.debug('cancelling existing timer...');
-      clearTimeout(this.timer);
-      this.timer = null;
-    }
-
-    if (!this.queue.length) {
-      this.logger.debug('queue is empty, nothing to flush');
-      setImmediate(callback);
-      return;
-    }
+  async flush(callback: AnalyticsFlushCallback = () => {}): Promise<FlushResponse | void> {
+    return await new Promise(async (resolve, reject) => {
+      // check if earlier flush was pushed to queue
+      this.logger.debug('in flush');
+
+      if (!this.enable) {
+        setImmediate(callback);
+        resolve();
+        return;
+      }
 
-    const items = this.queue.splice(0, this.flushAt);
-    const callbacks = items.map((item) => item.callback);
-    const messages = items.map((item) => {
-      // if someone mangles directly with queue
-      if (typeof item.message == 'object') {
-        item.message.sentAt = new Date();
+      if (this.timer) {
+        this.logger.debug('cancelling existing timer...');
+        clearTimeout(this.timer);
+        this.timer = null;
       }
-      return item.message;
-    });
 
-    const data = {
-      batch: messages,
-      sentAt: new Date(),
-    };
-    this.logger.debug('batch size is ' + items.length);
-    this.logger.trace('===data===', data);
+      if (!this.queue.length) {
+        this.logger.debug('queue is empty, nothing to flush');
+        setImmediate(callback);
+        resolve();
```

Comment:
could potentially resolve with an empty `FlushResponse` to consolidate the method return to a promise of `FlushResponse`. don't think it will help folks avoid a null check though, so i didn't include it. any pref?

Reply:
the empty flush response seems better. masking null checks is usually unhelpful but in this case i don't see a case where it's useful to return null if the queue happens to be empty